app/Http/Controllers/Playlist/PlaylistsController.php

```diff
@@ -564,6 +565,131 @@ public function completeItem(Request $request, $item_id)
         ]);
     }
 
+    /**
+     *
+     * @OA\Get(
+     *     path="/playlists/{playlist_id}/translate",
+     *     tags={"Playlists"},
+     *     summary="Translate a user's playlist",
+     *     operationId="v4_playlists.translate",
+     *     security={{"api_token":{}}},
+     *     @OA\Parameter(
+     *          name="playlist_id",
+     *          in="path",
+     *          required=true,
+     *          @OA\Schema(ref="#/components/schemas/Playlist/properties/id"),
+     *          description="The playlist id"
+     *     ),
+     *     @OA\Parameter(
+     *          name="bible_id",
+     *          in="query",
+     *          required=true,
+     *          @OA\Schema(ref="#/components/schemas/Bible/properties/id"),
+     *          description="The id of the bible that will be used to translate the playlist"
+     *     ),
+     *     @OA\Response(response=200, ref="#/components/responses/playlist")
+     * )
+     *
+     * @param $playlist_id
+     *
+     * @return mixed
+     *
+     *
+     */
+    public function translate(Request $request, $playlist_id)
+    {
+        $user = $request->user();
+
+        // Validate Project / User Connection
+        if (!empty($user) && !$this->compareProjects($user->id, $this->key)) {
+            return $this->setStatusCode(401)->replyWithError(trans('api.projects_users_not_connected'));
+        }
+
+        $bible_id = checkParam('bible_id', true);
+        $bible = Bible::whereId($bible_id)->first();
+
+        if (!$bible) {
+            return $this->setStatusCode(404)->replyWithError('Bible Not Found');
+        }
+
+        $playlist = $this->getPlaylist($user, $playlist_id);
+        if (!$playlist) {
+            return $this->setStatusCode(404)->replyWithError('Playlist Not Found');
+        }
+
+
+        $audio_fileset_types = collect(['audio_drama_stream', 'audio_stream', 'audio_drama', 'audio']);
+        $bible_audio_filesets = $bible->filesets->whereIn('set_type_code', $audio_fileset_types);
+
+        $translated_items = [];
+
+        foreach ($playlist->items as $item) {
+            $preferred_fileset = $audio_fileset_types->map(function ($type) use ($bible_audio_filesets, $item) {
+                return $this->getFileset($bible_audio_filesets, $type, $item->fileset->set_size_code);
```

Comment:
hi @mgonzalezs.  what does this logic do?  i'm thinking ideal case would be, for each item, use same fileset_type as the original if available, otherwise the first available in the list on line 621.  is that what this is doing?

Reply:
Now we look for the same fileset_type first then the preferred order of types as you suggested

Reply:
thanks!